File: request-manager/src/main/java/fii/request/manager/repository/WorkflowStepRepository.java
package fii.request.manager.repository;

import fii.request.manager.domain.WorkflowStep;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface WorkflowStepRepository extends JpaRepository<WorkflowStep, Long> {
    List<WorkflowStep> findByWorkflowId(Long workflowId);

    @Query("SELECT ws FROM WorkflowStep ws LEFT JOIN ws.workflow w WHERE w.id=:workflowId")
    List<WorkflowStep> findByWorkflowIdNotFetchingChildren(@Param(value="workflowId") Long workflowId);
}

File: workflow-manager-frontend/app/src/main/java/com/example/workflow_manager_frontend/data/repository/WorkflowStepRepository.kt
package com.example.workflow_manager_frontend.data.repository

import com.example.workflow_manager_frontend.domain.WorkflowStep

interface WorkflowStepRepository {
    suspend fun getWorkflowSteps(workflowId: Int?, fetchFromRemote: Boolean) : List<WorkflowStep>?

    suspend fun createWorkflowStep(workflowId: Int?, step: WorkflowStep)

    suspend fun updateWorkflowStep(workflowId: Int?, step: WorkflowStep)
}